feat: UI/UX for regions

Emyrk opened the pull request:
# What this does

When you select a workspace proxy, that proxy will be used for links on the workspace dashboard. App links change based on the selected proxy. The terminal changes what websocket url it uses based on the proxy.

Removed getting `application host` from other parts in the codebase. Using a context instead.

## Missing funcationality

Right now the terminal does not work with a workspace proxy because of the CSP policy. The error is:

```
rminalPage-b4b695c0.js:7 Refused to connect to ... because it violates the following Content Security Policy directive: "connect-src 'self' wss://localhost:3000 ws://localhost:3000".
```

This is known and not fixed in this PR. If you remove that CSP policy, it does work. We setup  CORs to work, but the CSP requires more refactoring. This will happen in another PR. See https://github.com/coder/coder/issues/7308

## Testing

Must have `moons` experiment enabled!

## Not done

If the selected proxy becomes unhealthy, the links will still direct to it.

# Styling TBD

All of this is feature flagged. We can make things prettier before we release. Just getting a functional UI in.

![Screenshot from 2023-04-25 14-50-35](https://user-images.githubusercontent.com/5446298/234387485-72f50e7f-bb00-4c60-9da2-f6087b2c8f28.png)

# Without experiment on

If `moons` experiment is not enabled, no UI changes happen.

![Screenshot from 2023-04-25 14-53-58](https://user-images.githubusercontent.com/5446298/234388173-c37b80dc-5d48-4726-aac7-c307e98b6f02.png)

Emyrk committed:
chore: Allow regular users to query for all workspaces

Emyrk committed:
Begin work on FE to add workspace proxy options to account settings

Emyrk committed:
Take origin file

Emyrk committed:
Remove excess diffs

Emyrk committed:
fixup! Remove excess diffs

Emyrk committed:
Update proxy page for regions endpoint

Emyrk committed:
Some basic selector for proxies

Emyrk committed:
Make hook for preferred proxy

Emyrk committed:
Make fmt

Emyrk committed:
Typo

deansheather commented on site/src/components/AppLink/AppLink.tsx:
```diff
@@ -29,6 +30,11 @@ export const AppLink: FC<AppLinkProps> = ({
+  const preferredProxy = usePreferredProxy()
+  const preferredPathBase = preferredProxy ? preferredProxy.path_app_url : ""
+  // Use the proxy host subdomain if it's configured.
+  appsHost = preferredProxy ? preferredProxy.wildcard_hostname : appsHost
```
The appsHost prop should probably be removed, as well as the apps host endpoint.

Emyrk replied:
I'm not sure I follow

deansheather replied:
See my comment about what the context should do below. Instead of this component calling the `/api/v2/applications/host` route to get the primary wildcard host it should only use the `/api/v2/regions` endpoint, and the old legacy route should be removed

Emyrk replied:
Will simplify this later.

deansheather commented on site/src/components/AppLink/AppLink.tsx:
```diff
@@ -29,6 +30,11 @@ export const AppLink: FC<AppLinkProps> = ({
   workspace,
   agent,
 }) => {
+  const preferredProxy = usePreferredProxy()
```
TBH this context should return `{ regions: [], selectedRegionID: "abcdef-...", preferredPathAppURL: "", preferredWildcardHostname: "" }`

The context can merge with the first region if they're not set on the selected one. This saves each consumer of this context from having to do the ternarys

Emyrk replied:
Good idea :+1:

Emyrk replied:
Fixed the ternaries

deansheather commented on site/src/hooks/usePreferredProxy.ts:
```diff
@@ -0,0 +1,20 @@
+    return undefined
+  }
+
+  const str = localStorage.getItem("preferred-proxy")
```
This localStorage should only store the ID, and it should get the regions list on startup (lazily perhaps)

Storing the whole JSON object means if any changes are made serverside they won't get considered on clients

Emyrk replied:
Yea, I am keeping the proxy for now for naming reasons.
The new context will fetch all the regions and update the proxy values if they are stale.

bpmct commented:
So an admin can set an icon for each workspace proxy?

Emyrk commented:
> So an admin can set an icon for each workspace proxy?

Yes. Figured that kinda our pattern for other resources and icons are fun

Emyrk committed:
Create workspace proxy context

Emyrk committed:
Use new context

Emyrk committed:
fixup! Use new context

Emyrk committed:
WorkspaceProxy context syncs with coderd on region responses

deansheather commented on enterprise/coderd/workspaceproxy.go:
```diff
@@ -156,6 +157,20 @@ func (api *API) postWorkspaceProxy(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	if strings.ToLower(req.Name) == "primary" {
```
cool!

deansheather commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,185 @@
+  // object. Use the preferred fields.
+  selectedRegion: Region | undefined
+  // PreferredPathAppURL is the URL of the proxy or it is the empty string
+  // to indicte using relative paths. To add a path to this:
```
```suggestion
  // to indicate using relative paths. To add a path to this:
```

deansheather commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,185 @@
+  selectedRegion: Region | undefined
+  // PreferredPathAppURL is the URL of the proxy or it is the empty string
+  // to indicte using relative paths. To add a path to this:
+  //  PreferredPathAppURL + "/path/to/app"
```
If this is the case we should make sure we strip the trailing slash when we populate the context below.

Emyrk replied:
I do!:
https://github.com/coder/coder/blob/stevenmasley%2Fregions/site/src/contexts/ProxyContext.tsx#L160

deansheather commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,185 @@
+  const { error: regionsError, isLoading: regionsLoading } = useQuery({
+    queryKey,
+    queryFn: getWorkspaceProxies,
+    // This onSucccess ensures the local storage is synchronized with the 
```
```suggestion
    // This onSuccess ensures the local storage is synchronized with the 
```

deansheather commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,185 @@
+/**
+ * ProxyProvider interacts with local storage to indicate the preferred workspace proxy.
+ */
+export const ProxyProvider: FC<PropsWithChildren> = ({ children }) => {
```
IMO this provider should lazily load all regions and have a `loading: bool` field. We can use the existing value from localStorage verbatim until we've loaded the regions list and can update it with a server-approved value.

Emyrk replied:
I don't feel a huge need to optimize. I removed the previous application host loading which was used in the workspace dashboard and terminal.

So optimizing this to lazy load would save a http call if the user would never hit the settings or workspace page.

I'd rather just be up front and make the call for now.

Emyrk committed:
Make fmt

Emyrk committed:
Move dashboard provider

Emyrk committed:
Fix authenticated providers

Emyrk committed:
Fix authenticated renders

Emyrk committed:
Merge remote-tracking branch 'origin/main' into stevenmasley/regions

Emyrk committed:
Make fmt

Emyrk committed:
Use auth render

Emyrk committed:
Fix terminal test render

Emyrk committed:
Make fmt

Emyrk committed:
Fix local storage load

Emyrk committed:
Fix terminals on the frontend to use proxies

Emyrk committed:
Remove CSP hole

Emyrk committed:
Add comment on origin patterns

deansheather commented on site/src/components/TerminalLink/TerminalLink.tsx:
```diff
@@ -3,6 +3,7 @@ import { SecondaryAgentButton } from "components/Resources/AgentButton"
 import { FC } from "react"
 import * as TypesGen from "../../api/typesGenerated"
 import { generateRandomString } from "../../utils/random"
+import { useProxy } from "contexts/ProxyContext"
```
```suggestion
```

Emyrk commented on site/src/AppRouter.tsx:
```diff
@@ -177,127 +188,145 @@ export const AppRouter: FC = () => {
           <Route element={<RequireAuth />}>
-            <Route element={<DashboardLayout />}>
-              <Route index element={<IndexPage />} />
+            <Route element={<AuthenticatedProviders />}>
```
This is the main route change. Added a provider to all authenticated routes.

BrunoQuaresma replied:
We should not need this, `<RequireAuth>` is exactly for that.

Emyrk commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,210 @@
+  }
+  // ******************************* //
+
+  // TODO: @emyrk Should make an api call to /regions endpoint to update the
+  // regions list.
```
Going to kick this to later.

Emyrk committed:
Add unit test for getURLs

BrunoQuaresma commented on site/src/AppRouter.tsx:
```diff
@@ -309,3 +338,14 @@ export const AppRouter: FC = () => {
+      </ProxyProvider>
+    </DashboardProvider>
+  )
+}
```
IMO, having a `RequireAuth` and `AuthenticatedProviders` is kinda confusing because RequireAuth should do that. I think the best would be to move the ProxyProvider inside the DashboardProvider and wrap the component there or even merge the DashboardProvider with ProxyProvider if that is only used inside the Dashboard.

Emyrk replied:
It is also used in the Terminal Page, and the settings page

BrunoQuaresma reviewed (commented):
Just a few things:
- Having `RequireAuth` and `AuthenticatedProviders` sounds redundant to me. I'm ok with keeping one of them only. 
- I see there are a lot of TODOS and import comments, I think we should address at least the comments on the imports. Not sure what about the TODOs... should we have tickets for them?

Emyrk committed:
remove some TODOs

Emyrk commented:
> Just a few things:
> 
> * Having `RequireAuth` and `AuthenticatedProviders` sounds redundant to me. I'm ok with keeping one of them only.
> * I see there are a lot of TODOS and import comments, I think we should address at least the comments on the imports. Not sure what about the TODOs... should we have tickets for them?

Removed some TODOs my bad.

The remaining TODOS:
- Style TODO. 
  - This feature cannot be seen yet, so rather not spin cycles in the PR making it pretty.
- Regions behavior on load. 
  - Currently the behavior is the user selects their workspace proxy. In future, we plan to do this automatically. So right now I don't want to optimize for the edge cases, but left a note for when I come back. Essentially the manual behavior now will become automatic, so any solutions we come up with will be discarded when we change the method of setting a proxy.

Emyrk committed:
Add another store

Emyrk committed:
Update site/src/components/TerminalLink/TerminalLink.tsx

Co-authored-by: Dean Sheather <dean@deansheather.com>

Emyrk committed:
Fix stories

Emyrk committed:
Move providers into requrie auth

Emyrk committed:
Fix imports

Emyrk committed:
Fix 2 stories

Emyrk committed:
Stories did not have subdomains on

Emyrk committed:
Merge remote-tracking branch 'origin/main' into stevenmasley/regions

Emyrk committed:
Fmt after merge

Emyrk committed:
Fix port forward story

Kira-Pilot commented:
I'm seeing this panic in the server logs when I run your branch. I haven't seen it on `main` (although maybe I missed it?):
````
[API] 2023-04-27 20:33:42.086 [WARN]    (coderd)        <./coderd/httpmw/recover.go:25> Recover.func1.1.1       panic serving http request (recovered)  {"panic": "runtime error: invalid memory address or nil pointer dereference"} ...
[API] "stack": goroutine 1751 [running]:
[API]          runtime/debug.Stack()
[API]           /usr/local/go/src/runtime/debug/stack.go:24 +0x65
[API]          github.com/coder/coder/coderd/httpmw.Recover.func1.1.1()
[API]           /home/coder/coder/coderd/httpmw/recover.go:28 +0xc8
[API]          panic({0x222f420, 0x191346d0})
[API]           /usr/local/go/src/runtime/panic.go:884 +0x213
[API]          go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End.func1()
[API]           /home/coder/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.14.0/trace/span.go:383 +0x2a
[API]          go.opentelemetry.io/otel/sdk/trace.(*recordingSpan).End(0xc01feab680, {0x0, 0x0, 0x24?})
[API]           /home/coder/go/pkg/mod/go.opentelemetry.io/otel/sdk@v1.14.0/trace/span.go:415 +0x9fe
[API]          panic({0x222f420, 0x191346d0})
[API]           /usr/local/go/src/runtime/panic.go:884 +0x213
[API]          github.com/coder/coder/enterprise/coderd/proxyhealth.(*ProxyHealth).HealthStatus(...)
[API]           /home/coder/coder/enterprise/coderd/proxyhealth/proxyhealth.go:159
[API]          github.com/coder/coder/enterprise/coderd.(*API).regions(0xc011c27a40, {0x17e1b620, 0xc01feb0940}, 0x0?)
[API]           /home/coder/coder/enterprise/coderd/workspaceproxy.go:61 +0x247
[API]          net/http.HandlerFunc.ServeHTTP(0xc01f924ae0?, {0x17e1b620?, 0xc01feb0940?}, 0xc01fea75a8?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/coder/coder/coderd/httpmw.ExtractAPIKeyMW.func1.1({0x17e1b620, 0xc01feb0940}, 0xc01fecc100)
[API]           /home/coder/coder/coderd/httpmw/apikey.go:133 +0x2dc
[API]          net/http.HandlerFunc.ServeHTTP(0xc0011f1a40?, {0x17e1b620?, 0xc01feb0940?}, 0xc012646310?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/go-chi/chi/v5.(*ChainHandler).ServeHTTP(0x2212240?, {0x17e1b620?, 0xc01feb0940?}, 0xc01fe19a60?)
[API]           /home/coder/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/chain.go:31 +0x2c
[API]          github.com/go-chi/chi/v5.(*Mux).routeHTTP(0xc0013a6c60, {0x17e1b620, 0xc01feb0940}, 0xc01fecc100)
[API]           /home/coder/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:442 +0x216
[API]          net/http.HandlerFunc.ServeHTTP(0xc0011f1110?, {0x17e1b620?, 0xc01feb0940?}, 0x15?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/go-chi/httprate.(*rateLimiter).Handler.func1({0x17e1b620, 0xc01feb0940}, 0xc01fe19a61?)
[API]           /home/coder/go/pkg/mod/github.com/go-chi/httprate@v0.7.1/limiter.go:124 +0x9ad
[API]          net/http.HandlerFunc.ServeHTTP(0xc01199a4c8?, {0x17e1b620?, 0xc01feb0940?}, 0xc01199a4a8?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/go-chi/chi/v5.(*Mux).ServeHTTP(0xc0013a6c60, {0x17e1b620, 0xc01feb0940}, 0xc01fecc100)
[API]           /home/coder/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:71 +0x355
[API]          github.com/go-chi/chi/v5.(*Mux).Mount.func1({0x17e1b620, 0xc01feb0940}, 0xc01fecc100)
[API]           /home/coder/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:314 +0x1c4
[API]          net/http.HandlerFunc.ServeHTTP(0x2212240?, {0x17e1b620?, 0xc01feb0940?}, 0xc01f708d04?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/go-chi/chi/v5.(*Mux).routeHTTP(0xc0013a6720, {0x17e1b620, 0xc01feb0940}, 0xc01fecc100)
[API]           /home/coder/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:442 +0x216
[API]          net/http.HandlerFunc.ServeHTTP(0xc0001ae000?, {0x17e1b620?, 0xc01feb0940?}, 0x21?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/justinas/nosurf.(*CSRFHandler).handleSuccess(...)
[API]           /home/coder/go/pkg/mod/github.com/justinas/nosurf@v1.1.1/handler.go:187
[API]          github.com/justinas/nosurf.(*CSRFHandler).ServeHTTP(0xc011afc480, {0x17e1b620, 0xc01feb0940}, 0x16?)
[API]           /home/coder/go/pkg/mod/github.com/justinas/nosurf@v1.1.1/handler.go:144 +0x634
[API]          github.com/coder/coder/coderd.New.func4.1({0x17e1b620, 0xc01feb0940}, 0xc01fea2ea0?)
[API]           /home/coder/coder/coderd/coderd.go:422 +0x128
[API]          net/http.HandlerFunc.ServeHTTP(0x0?, {0x17e1b620?, 0xc01feb0940?}, 0x15?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/coder/coder/coderd.New.func3.1({0x17e1b620, 0xc01feb0940}, 0x0?)
[API]           /home/coder/coder/coderd/coderd.go:412 +0x133
[API]          net/http.HandlerFunc.ServeHTTP(0x0?, {0x17e1b620?, 0xc01feb0940?}, 0x0?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/coder/coder/coderd/workspaceapps.(*Server).HandleSubdomain.func1.1({0x17e1b620?, 0xc01feb0940?}, 0xc01feb2b01?)
[API]           /home/coder/coder/coderd/workspaceapps/proxy.go:336 +0x97
[API]          net/http.HandlerFunc.ServeHTTP(0x0?, {0x17e1b620?, 0xc01feb0940?}, 0x14f9c93d48?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/coder/coder/coderd/httpmw.Prometheus.func1.1({0x17e1b620?, 0xc01feb0940}, 0xc01fecc000)
[API]           /home/coder/coder/coderd/httpmw/prometheus.go:90 +0x2f2
[API]          net/http.HandlerFunc.ServeHTTP(0xc01fe19a50?, {0x17e1b620?, 0xc01feb0940?}, 0xc011589298?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/coder/coder/coderd/httpmw.Logger.func1.1({0x17e1b620?, 0xc01feb0940?}, 0xc01fecc000)
[API]           /home/coder/coder/coderd/httpmw/logger.go:31 +0x41d
[API]          net/http.HandlerFunc.ServeHTTP(0xc01fe19a40?, {0x17e1b620?, 0xc01feb0940?}, 0x191e1f10?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/coder/coder/coderd/httpmw.ExtractRealIP.func1.1({0x17e1b620, 0xc01feb0940}, 0xc01f745f00)
[API]           /home/coder/coder/coderd/httpmw/realip.go:50 +0x21d
[API]          net/http.HandlerFunc.ServeHTTP(0x17e22910?, {0x17e1b620?, 0xc01feb0940?}, 0x191e1f10?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/coder/coder/coderd/httpmw.AttachAuthzCache.func1({0x17e1b620, 0xc01feb0940}, 0xc01f745e00)
[API]           /home/coder/coder/coderd/httpmw/authz.go:49 +0x16e
[API]          net/http.HandlerFunc.ServeHTTP(0x23e76c0?, {0x17e1b620?, 0xc01feb0940?}, 0x12?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/coder/coder/coderd/httpmw.AttachRequestID.func1({0x17e1b620, 0xc01feb0940}, 0xc01f745d00)
[API]           /home/coder/coder/coderd/httpmw/requestid.go:33 +0x39c
[API]          net/http.HandlerFunc.ServeHTTP(0x23e76c0?, {0x17e1b620?, 0xc01feb0940?}, 0x17e1fcd0?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/coder/coder/coderd/tracing.Middleware.func1.1({0x17e1b620?, 0xc01feb0940?}, 0xc01f745c00)
[API]           /home/coder/coder/coderd/tracing/httpmw.go:71 +0x682
[API]          net/http.HandlerFunc.ServeHTTP(0xc0117849d0?, {0x17e1b620?, 0xc01feb0940?}, 0xa?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/coder/coder/coderd/tracing.StatusWriterMiddleware.func1({0x17e1ef20?, 0xc01d9cf880}, 0xe0?)
[API]           /home/coder/coder/coderd/tracing/status_writer.go:31 +0x93
[API]          net/http.HandlerFunc.ServeHTTP(0xc0117849f8?, {0x17e1ef20?, 0xc01d9cf880?}, 0x30?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/coder/coder/coderd/httpmw.Recover.func1.1({0x17e1ef20?, 0xc01d9cf880?}, 0x19130b01?)
[API]           /home/coder/coder/coderd/httpmw/recover.go:44 +0xd0
[API]          net/http.HandlerFunc.ServeHTTP(0x17e22868?, {0x17e1ef20?, 0xc01d9cf880?}, 0x19130be0?)
[API]           /usr/local/go/src/net/http/server.go:2122 +0x2f
[API]          github.com/go-chi/chi/v5.(*Mux).ServeHTTP(0xc0013a6720, {0x17e1ef20, 0xc01d9cf880}, 0xc01fe8e000)
[API]           /home/coder/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:88 +0x310
[API]          net/http.serverHandler.ServeHTTP({0xc01fe8c060?}, {0x17e1ef20, 0xc01d9cf880}, 0xc01fe8e000)
[API]           /usr/local/go/src/net/http/server.go:2936 +0x316
[API]          net/http.(*conn).serve(0xc01fe9e360, {0x17e22910, 0xc001090300})
[API]           /usr/local/go/src/net/http/server.go:1995 +0x612
[API]          created by net/http.(*Server).Serve
[API]           /usr/local/go/src/net/http/server.go:3089 +0x5ed
````

Kira-Pilot commented on site/src/pages/UserSettingsPage/WorkspaceProxyPage/WorkspaceProxyPage.tsx:
```diff
@@ -0,0 +1,77 @@
+import { useWorkspaceProxiesData } from "./hooks"
+import { displayError } from "components/GlobalSnackbar/utils"
+import { useProxy } from "contexts/ProxyContext"
+// import { ConfirmDeleteDialog } from "./components"
+// import { Stack } from "components/Stack/Stack"
+// import Button from "@material-ui/core/Button"
+// import { Link as RouterLink } from "react-router-dom"
+// import AddIcon from "@material-ui/icons/AddOutlined"
+// import { APIKeyWithOwner } from "api/typesGenerated"
```
```suggestion
```

Kira-Pilot commented on site/src/pages/UserSettingsPage/WorkspaceProxyPage/WorkspaceProxyPage.tsx:
```diff
@@ -0,0 +1,77 @@
+      workspace. To get the best experience, choose the workspace proxy that is
+      closest located to you.
+    </Trans>
+  )
```
We don't need a `Trans` component here. You can just use this text directly or assign it a `const` if you feel like it's cluttering up the JSX.

Kira-Pilot commented on site/src/pages/UserSettingsPage/WorkspaceProxyPage/hooks.ts:
```diff
@@ -0,0 +1,15 @@
+// Loads all workspace proxies
+export const useWorkspaceProxiesData = () => {
+  const queryKey = ["workspace-proxies"]
+  const result = useQuery({
+    queryKey,
+    queryFn: () => getWorkspaceProxies(),
+  })
+
+  return {
+    ...result,
+  }
```
```suggestion
  return useQuery({
    queryKey,
    queryFn: getWorkspaceProxies,
  })
```

Kira-Pilot commented on site/src/api/api.ts:
```diff
@@ -897,6 +897,15 @@ export const getFile = async (fileId: string): Promise<ArrayBuffer> => {
 
+export const getWorkspaceProxies =
+  async (): Promise<TypesGen.RegionsResponse> => {
+    const response = await axios.get<TypesGen.RegionsResponse>(
+      `/api/v2/regions`,
+      {},
+    )
```
```suggestion
    const response = await axios.get<TypesGen.RegionsResponse>(
      `/api/v2/regions`
    )
```

Kira-Pilot commented on site/src/pages/UserSettingsPage/WorkspaceProxyPage/WorkspaceProxyPage.tsx:
```diff
@@ -0,0 +1,77 @@
+  } = useWorkspaceProxiesData()
+
+  return (
+    <>
```
We don't need to wrap this component return in a JSX fragment since the `<Section />` element already wraps all the children.

Kira-Pilot commented on site/src/pages/UserSettingsPage/WorkspaceProxyPage/WorkspaceProxyPage.tsx:
```diff
@@ -0,0 +1,77 @@
+  const { proxy, setProxy } = useProxy()
+
+  const {
+    data: response,
```
NIT: idk that naming this to "response" lends a lot of clarity.

Kira-Pilot commented on site/src/pages/UserSettingsPage/WorkspaceProxyPage/WorkspaceProxyPage.tsx:
```diff
@@ -0,0 +1,77 @@
+        layout="fluid"
+      >
+        <WorkspaceProxyPageView
+          proxies={response ? response.regions : []}
```
The proxies prop can be undefined so you can simplify to:
```suggestion
          proxies={response?.regions}
```

Kira-Pilot commented on site/src/pages/UserSettingsPage/WorkspaceProxyPage/WorkspaceProxyView.tsx:
```diff
@@ -0,0 +1,80 @@
+  selectProxyError?: Error | unknown
+}
+
+export const WorkspaceProxyPageView: FC<
```
Can we have the file name match the default export name?

Emyrk committed:
ProxyPageView -> ProxyView

Emyrk committed:
PR comment cleanup

Emyrk committed:
Fix moon feature flag panic

Emyrk committed:
Make fmt

Emyrk commented:
> I'm seeing this panic in the server logs when I run your branch. I haven't seen it on `main` (although maybe I missed it?):

Fixed, good catch. We added a route that was gated by the feature flag. We changed the endpoint to always be accessible, and forgot to update the code to reflect that.

I fixed this.

Emyrk committed:
Fix typo

Kira-Pilot commented:
Throughout this PR, seems like we're using 'proxy' and 'region' interchangeably. Can we pick one and be consistent?

Kira-Pilot commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,211 @@
+  const queryKey = ["get-regions"]
+  const { error: regionsError, isLoading: regionsLoading } = useQuery({
+    queryKey,
+    queryFn: getWorkspaceProxies,
```
Because this Provider wraps the whole app, we are making this call everywhere, including when we don't have this feature enabled which results in an automatic 500 console error:
![Screenshot 2023-04-27 at 7 47 06 PM](https://user-images.githubusercontent.com/19142439/235013656-057b2583-c789-4b32-848f-6e563c05b429.png)
We should figure out another strategy here (maybe returning early after an entitlement check) so that we don't have persistent console errors.

Emyrk replied:
Did you pull the latest?

I made a change so the `/regions` endpoint is **always available** to fix this. Even if the feature is not enabled.

Kira-Pilot commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,211 @@
+  window.localStorage.setItem("preferred-proxy", JSON.stringify(saved))
+}
+
+export const loadPreferredProxy = (): PreferredProxy | undefined => {
```
```suggestion
const loadPreferredProxy = (): PreferredProxy | undefined => {
```

Kira-Pilot commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,211 @@
+
+export const loadPreferredProxy = (): PreferredProxy | undefined => {
+  const str = localStorage.getItem("preferred-proxy")
+  if (str === undefined || str === null) {
```
[It looks to me](https://developer.mozilla.org/en-US/docs/Web/API/Storage/getItem) like this is either going to return a value, or else it's going to return `null`, so I think we can simplify:

```suggestion
  if (!str) {
```

Kira-Pilot commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,211 @@
+}
+
+export const loadPreferredProxy = (): PreferredProxy | undefined => {
+  const str = localStorage.getItem("preferred-proxy")
+  if (str === undefined || str === null) {
+    return undefined
+  }
+  const proxy: PreferredProxy = JSON.parse(str)
+  if (proxy.selectedRegion === undefined || proxy.selectedRegion === null) {
+    return undefined
+  }
+  return proxy
```
[It looks to me](https://developer.mozilla.org/en-US/docs/Web/API/Storage/getItem) like this is either going to return a value, or else it's going to return `null`, so I think we can simplify:

```suggestion
  const str = localStorage.getItem("preferred-proxy")
  if (!str) {
    return undefined
  }

  return JSON.parse(str)
```

Kira-Pilot commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,211 @@
+    savedProxy = getURLs([])
+  }
+
+  // The initial state is no regions and no selected region.
```
Is this comment accurate? Seems like the initial state is either a) the saved region (proxy?) from local storage or b) the default region (proxy?).

Emyrk replied:
I will fix the comment

Kira-Pilot commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,211 @@
+  const [proxy, setProxy] = useState<PreferredProxy>(savedProxy)
+  const setAndSaveProxy = (
+    regions: Region[],
+    selectedRegion: Region | undefined,
```
`setAndSaveProxy` should only ever be called deliberately, when we have a region (proxy?) to set. You can check this in your type for `ProxyRow` in `WorkspaceProxyRow.tsx` - the proxy argument is not optional.
```suggestion
    selectedRegion: Region,
```

Kira-Pilot replied:
What do you think about not passing `regions` in as an arg, considering we load it right below?
I left another comment in your`WorkspaceProxyPage/hooks.ts` file -  I think we can get away without that hook, considering we make the same API call in here. Should simplify a fair amount.

Emyrk replied:
I'm not sure what you mean not passing in the regions.

Kira-Pilot commented on site/src/pages/UserSettingsPage/WorkspaceProxyPage/hooks.ts:
```diff
@@ -0,0 +1,11 @@
+import { useQuery } from "@tanstack/react-query"
+import { getWorkspaceProxies } from "api/api"
+
+// Loads all workspace proxies
```
I don't think we need this secondary hook. Instead, we should roll whatever you need out of this into your `ProxyContext`. We already make a call in there to `getWorkspaceProxies` .

Emyrk replied:
You are right.

Kira-Pilot commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,211 @@
+  // If the experiment is disabled, then make the setState do a noop.
+  // This preserves an empty state, which is the default behavior.
+  if (!dashboard?.experiments.includes("moons")) {
+    const value = getURLs([])
```
Instead of defining a completely different `<ProxyContent.Provider />` here, we can use [our `createContext` hook](https://github.com/coder/coder/pull/7283/files#diff-1beec2c2724792e5a29bc182e07983cd672b52ea138eca517ddb65c5e30bb500R34) to define initial values, e.g:
````
export const ProxyContext = createContext<ProxyContextValue>({
  proxy: {...your default}
  isLoading: false
  setProxy: () => {}
})
````
Then we can bail out with an early return if the experiment isn't turned on. We want to avoid returning alternative `<ProxyContent.Provider />` elements.

Emyrk replied:
I'm not sure how to do this :thinking:. I need to set all the values except for the `SetProxy` function

Emyrk replied:
Fixed in chat with Kira

Kira-Pilot commented on site/src/contexts/ProxyContext.tsx:
```diff
@@ -0,0 +1,211 @@
+ * @param regions Is the list of regions returned by coderd. If this is empty, default behavior is used.
+ * @param selectedRegion Is the region the user has selected. If this is undefined, default behavior is used.
+ */
+export const getURLs = (
```
This isn't really just getting URLs, is it? It's really setting the entire PreferredProxy object.

Emyrk replied:
Updated the name

Emyrk committed:
Rename getUrls

Emyrk committed:
Rename regions to proxies

Emyrk committed:
Only do 1 api call based on experiment

Emyrk committed:
Cleanup args to take just the selected proxy

Delete extra hook
Renames regions -> proxies

Emyrk committed:
Renames regions -> proxies

Emyrk committed:
Fix stories

Emyrk committed:
Move funciton back to bottom

Emyrk committed:
Fix onSuccess of proxy provider

Emyrk committed:
Make fmt

Kira-Pilot commented on site/src/components/Resources/AgentRow.tsx:
```diff
@@ -249,15 +248,16 @@ export const AgentRow: FC<AgentRowProps> = ({
-                username={workspace.owner_name}
-              />
-            )}
+            {proxy.preferredWildcardHostname !== undefined &&
+              proxy.preferredWildcardHostname !== "" && (
```
Can we just
```suggestion
            {proxy.preferredWildcardHostname && (
```

Emyrk replied:
I will always forget this lol

Kira-Pilot commented on site/src/pages/UserSettingsPage/WorkspaceProxyPage/WorkspaceProxyRow.tsx:
```diff
@@ -0,0 +1,81 @@
+      data-testid={`${proxy.name}`}
+      {...clickable}
+      // Make sure to include our classname here.
+      className={combineClasses(classes)}
```
Here's a cleaner way to use conditional classes:
```suggestion
      className={combineClasses({
          [clickable.className]: true,
          [styles.preferredrow]: preferred,
        })}
```

Emyrk replied:
TIL

Kira-Pilot commented on site/src/pages/UserSettingsPage/WorkspaceProxyPage/WorkspaceProxyRow.tsx:
```diff
@@ -0,0 +1,81 @@
+    onSelectRegion(proxy)
+  })
+
+  const classes = [clickable.className]
+
+  if (preferred) {
+    classes.push(styles.preferredrow)
+  }
```
```suggestion
```

Kira-Pilot reviewed (approved):
Nice work here!

Emyrk committed:
Simplify some ts
